src/compiler/crystal/semantic/main_visitor.cr

```diff
@@ -2139,8 +2139,15 @@ module Crystal
 
           # If the loop is endless
           if endless
-            after_while_var.bind_to(while_var)
-            after_while_var.nil_if_read = while_var.nil_if_read?
+            # If a variable was re-assigned before the first break expression,
+            # the type at the end of the while body is inaccessible upon exit
+            # because the assignment must have executed before the next chance
+            # to break
```

Comment:
Would you mind adding a comment code snippet to describe this situation?

I'm pretty sure this PR is correct (as usual from you), I just want to make sure in the future we still remember what this means. I did a poor job of not adding code snippets to explain the code better, so I want to improve that.